zerver/decorator.py

```diff
@@ -681,32 +698,30 @@ def wrapper(func):
         def wrapped_func(request, *args, **kwargs):
             # type: (HttpRequest, *Any, **Any) -> HttpResponse
 
-            # It is really tempting to not even wrap our original function
-            # when settings.RATE_LIMITING is False, but it would make
-            # for awkward unit testing in some situations.
-            if not settings.RATE_LIMITING:
-                return func(request, *args, **kwargs)
-
             if client_is_exempt_from_rate_limiting(request):
                 return func(request, *args, **kwargs)
 
-            try:
-                user = request.user
-            except Exception:
-                # TODO: This logic is not tested, and I'm not sure we are
-                # doing the right thing here.
-                user = None
-
-            if not user:
-                logging.error("Requested rate-limiting on %s but user is not authenticated!" %
-                              func.__name__)
-                return func(request, *args, **kwargs)
-
-            # Rate-limiting data is stored in redis
-            # We also only support rate-limiting authenticated
-            # views right now.
-            # TODO(leo) - implement per-IP non-authed rate limiting
-            rate_limit_user(request, user, domain)
+            if settings.RATE_LIMITING:
+                # User based rate limiting. Works when a user is logged in.
+                try:
+                    user = request.user
+                except Exception:
+                    # Not authed. Ignore and move to IP based rate limiting.
+                    logging.debug("Requested rate-limiting on {} but user is "
+                                  "not authenticated!".format(func.__name__))
+                else:
+                    # This function can raise RateLimited exception which is
+                    # processed by
+                    # zerver.middleware.RateLimitMiddleware.process_exception.
+                    rate_limit_user(request, user, domain)
+
+            if settings.IP_RATE_LIMITING:
+                ip = get_ip(request)
```

Comment:
It just checks for the IP in these headers, details [here](https://github.com/un33k/django-ipware/blob/master/ipware/defaults.py):
```
'HTTP_X_FORWARDED_FOR', 'X_FORWARDED_FOR',  # (client, proxy1, proxy2) OR (proxy2, proxy1, client)
        'HTTP_CLIENT_IP',
        'HTTP_X_REAL_IP',
        'HTTP_X_FORWARDED',
        'HTTP_X_CLUSTER_CLIENT_IP',
        'HTTP_FORWARDED_FOR',
        'HTTP_FORWARDED',
        'HTTP_VIA',
'REMOTE_ADDR',
```

It basically just have one function:
```


def get_ip(request, real_ip_only=False, right_most_proxy=False):
    """
    Returns client's best-matched ip-address, or None
    """
    best_matched_ip = None
    for key in defs.IPWARE_META_PRECEDENCE_ORDER:
        value = request.META.get(key, request.META.get(key.replace('_', '-'), '')).strip()
        if value is not None and value != '':
            ips = [ip.strip().lower() for ip in value.split(',')]
            if right_most_proxy and len(ips) > 1:
                ips = reversed(ips)
            for ip_str in ips:
                if ip_str and is_valid_ip(ip_str):
                    if not ip_str.startswith(NON_PUBLIC_IP_PREFIX):
                        return ip_str
                    if not real_ip_only:
                        loopback = defs.IPWARE_LOOPBACK_PREFIX
                        if best_matched_ip is None:
                            best_matched_ip = ip_str
                        elif best_matched_ip.startswith(loopback) and not ip_str.startswith(loopback):
                            best_matched_ip = ip_str
return best_matched_ip
```